lib/resty/ldap/filter.lua

```diff
@@ -0,0 +1,146 @@
+local lpeg = require('lpeg')
+local P, R, S, V = lpeg.P, lpeg.R, lpeg.S, lpeg.V
+local C, Ct, Cmt, Cg, Cc, Cf, Cmt = lpeg.C, lpeg.Ct, lpeg.Cmt, lpeg.Cg, lpeg.Cc, lpeg.Cf, lpeg.Cmt
+
+local string_char = string.char
+local table_insert = table.insert
+local table_concat = table.concat
+
+local _M = {}
+
+-- Const
+_M.OP_TYPE_AND = 'and'
+_M.OP_TYPE_OR = 'or'
+_M.OP_TYPE_NOT = 'not'
+_M.ITEM_TYPE_SUBSTRING = 'substring'
+_M.ITEM_TYPE_SIMPLE = 'simple'
+_M.ITEM_TYPE_PRESENT = 'present'
+_M.FILTER_TYPE_EQUAL = 'equal'
+_M.FILTER_TYPE_APPROX = 'approx'
+_M.FILTER_TYPE_GREATER = 'greater'
+_M.FILTER_TYPE_LESS = 'less'
+
+
+local function pack(...)
+    return { n = select('#', ...), ... }
+end
+
+-- Utility
+local _= V
+
+local function maybe(pattern)
+    if type(pattern) == 'string' then pattern = V(pattern) end
+    return pattern ^ -1
+end
+
+local function list(pattern, min)
+    if type(pattern) == 'string' then pattern = V(pattern) end
+    return Ct((pattern) ^ (min or 0))
+end
+
+-- Formatters
+local cOPBody = function(op_type, ...)
+    return {
+        op_type = op_type,
+        items = (...)[1] ~= nil and (...) or {...}
+    }
+end
+
+local cItemBody = function(item_type, ...)
+    local args = {...}
+    return {
+        item_type = item_type,
+        attribute_description = args[1].attribute_description,
+        filter_type = args[2].filter_type,
+        attribute_value = args[3].attribute_value,
+    }
+end
+
+-- Simple types
+local rawValue = (P'_' + R('az', 'AZ')) * (P'_' + R '09' + R('az', 'AZ')) ^ 0
+
+-- Grammar
+local filter = P{
+    _'FILTER' / function(f)
```

Comment:
it is not easy to understand `_`, it is a function.

in Lua land, the variable can be ignore when we use `_`, here is an example:

```lua
local _, err = fun(... ...)
... ...
```

Reply:
ditto